### PR TITLE
nrf_desktop: dfu: Config channel should distinguish bootloader type

### PR DESCRIPTION
**Pull request by zycz**
Add fetch of bootloader variant in dfu module
Check if the bootloader file passed is compatible while doing dfu

Jira: NCSDK-12781

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

### CODE REVIEW

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: scripts/hid_configurator/configurator_cli.py [hunk 1]
```diff
@@ -59,6 +59,10 @@ def perform_dfu(dev, args):
         print('Cannot read image version from file')
         return
 
+    if info.bootloader_variant != img_file.bootloader_name:
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply scripts/hid_configurator/configurator_cli.py:62]: We could also add the check to the `DfuImage._parse_dfu_image_bin_path`. Currently we verify there among others if the image is meant to be used for given board.

File: applications/nrf_desktop/src/modules/dfu.c [hunk 2]
```diff
@@ -94,7 +97,8 @@ const static char * const opt_descr[] = {
 	[DFU_OPT_DATA] = "data",
 	[DFU_OPT_SYNC] = "sync",
 	[DFU_OPT_REBOOT] = "reboot",
-	[DFU_OPT_FWINFO] = "fwinfo"
+	[DFU_OPT_FWINFO] = "fwinfo",
+	[DFU_OPT_BOOTLOADER_VAR] = "bootloader_var"
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/dfu.c:101]: please use:
```
[DFU_OPT_VARIANT] = OPT_DESCR_MODULE_VARIANT,
```

There is no reason to introduce new name.

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: scripts/hid_configurator/modules/dfu.py [hunk 3]
```diff
@@ -339,6 +359,12 @@ def get_dfu_image_version(self):
         else:
             return self.bootloader_api['get_dfu_image_version'](self.image_bin_path)
 
+    def get_dfu_image_bootloader_var(self):
+        if self.image_bin_path is None:
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:363]: Why checking if bin path is not none? This seems unrelated

File: scripts/hid_configurator/modules/dfu.py [hunk 4]
```diff
@@ -347,10 +373,11 @@ def __del__(self):
 
 
 def fwinfo(dev):
-    success, fetched_data = dev.config_get('dfu', 'fwinfo')
+    success_data, fetched_data = dev.config_get('dfu', 'fwinfo')
+    success_bootloader, fetched_bootloader_variant = dev.config_get('dfu', 'bootloader_var')
 
-    if success and fetched_data:
-        fw_info = FwInfo(fetched_data)
+    if success_data and success_bootloader and fetched_data and fetched_bootloader_variant:
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:379]: btw. From this point, the Python scripts related to the DFU would not longer be compatible with previous firmware. Maybe it's worth to inform about it in docs or release notes ?

File: applications/nrf_desktop/src/modules/dfu.c (hunk 2 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/dfu.c:101]: That could give us more flexibility while adding bootloader-specific functionalities. On the other hand it would require bigger changes on the Python side (from this point we will need to refer to the module using either `dfu/mcuboot` or `dfu/b0`)

File: scripts/hid_configurator/modules/dfu.py (hunk 4 above)
Comment (zycz) [reply scripts/hid_configurator/modules/dfu.py:379]: Ok, will put information in release note.

File: applications/nrf_desktop/src/modules/dfu.c (hunk 2 above)
Comment (zycz) [reply applications/nrf_desktop/src/modules/dfu.c:101]: Ok we can introduce it. However there is a problem that right now that we cannot get variant name without knowing variant name (to fetch data you need to use command dfu/'variant_name'). I sugest to change current solution so that calling 
DEVICE config _module_ 
would return module variant.

File: scripts/hid_configurator/NrfHidDevice.py [hunk 5]
```diff
@@ -485,3 +485,9 @@ def config_get(self, module_name, option_name, poll_interval=POLL_INTERVAL_DEFAU
 
     def config_set(self, module_name, option_name, value, poll_interval=POLL_INTERVAL_DEFAULT):
         return self._config_operation(module_name, option_name, False, value, poll_interval)
+
+    def get_dfu_module_name(self):
```
Comment (pdunaj) [inline scripts/hid_configurator/NrfHidDevice.py:489]: can it be generic? maybe it will be useful for some other place
e.g.
```
get_module_name_by_root(self, name):
   for ket in self.dev_config:
      if '{}/'.format(name) in key:
         return key
   return None
```

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: scripts/hid_configurator/modules/dfu.py [hunk 6]
```diff
@@ -376,7 +397,9 @@ def fwreboot(dev):
 
 
 def dfu_sync(dev):
-    success, fetched_data = dev.config_get('dfu', 'sync')
+    dfu_module_name = dev.get_module_variant_name_by_root('dfu')
```
Comment (pdunaj) [inline scripts/hid_configurator/modules/dfu.py:400]: why returning only the variant suffix? we could return the whole name (that is "dfu/variant_name")
it seems to me that that would be easier to use

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: scripts/hid_configurator/modules/dfu.py (hunk 6 above)
Comment (zycz) [reply scripts/hid_configurator/modules/dfu.py:400]: We also use only variant_name for some purposes (compering with variant name during DFU). I thought it will be more readable that way.

File: scripts/hid_configurator/modules/dfu.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 import imgtool.image
 
 DFU_SYNC_INTERVAL = 1
-
+OPT_DESCR_MODULE_VARIANT = 'module_variant'
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:21]: Do we use it anywhere?

File: scripts/hid_configurator/NrfHidDevice.py [hunk 8]
```diff
@@ -485,3 +485,9 @@ def config_get(self, module_name, option_name, poll_interval=POLL_INTERVAL_DEFAU
 
     def config_set(self, module_name, option_name, value, poll_interval=POLL_INTERVAL_DEFAULT):
         return self._config_operation(module_name, option_name, False, value, poll_interval)
+
+    def get_module_variant_name_by_root(self, name):
```
Comment (MarekPieta) [inline scripts/hid_configurator/NrfHidDevice.py:489]: Maybe call it: `get_module_variant(self, name)`? That would be much simpler.

I would also prefer for this function to return whole name used to identify the module to avoid using information about naming convention in other DFU related scripts if it's not necessary (we keep information in a single place). So that we may call it e.g. `get_complete_module_name`?

File: scripts/hid_configurator/configurator_cli.py [hunk 9]
```diff
@@ -45,7 +45,7 @@ def perform_dfu(dev, args):
         return
     img_ver_dev = info.get_fw_version()
 
-    img_file = DfuImage(args.dfu_image, info, dev.get_board_name())
+    img_file = DfuImage(args.dfu_image, info, dev)
```
Comment (MarekPieta) [inline scripts/hid_configurator/configurator_cli.py:48]: Maybe we could only pass board name and bootloader name here? To avoid passing data that should not be used by the DfuImage class.

File: scripts/hid_configurator/modules/dfu.py (hunk 6 above)
Comment (MarekPieta) [reply scripts/hid_configurator/modules/dfu.py:400]: I think returning complete module name as @pdunaj suggested would be better approach here. Not only because it's simpler, but also it avoids constructing complete module name from name and variant in `dfu.py`. The `dfu.py` should not have information about naming format in `NrfHidDevice` class I think. It's not necessary here.

File: applications/nrf_desktop/doc/dfu.rst [hunk 10]
```diff
@@ -149,6 +150,11 @@ sync
 
    Fetching the ``sync`` option also triggers moving the image data from the RAM buffer to flash.
 
+.. _dfu_bootloader_var:
+
+bootloader variant
```
Comment (MarekPieta) [inline applications/nrf_desktop/doc/dfu.rst:155]: The option is actually named OPT_DESCR_MODULE_VARIANT ("module_variant") now. Please update.

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: scripts/hid_configurator/modules/dfu.py (hunk 6 above)
Comment (zycz) [reply scripts/hid_configurator/modules/dfu.py:400]: I guess I am overvoted. So I will change it.  For me it is counter intuitive - Would not understand what is complete module name while reading this function. I will add comment explaining what is " complete module name".
Only variant name is as simple as with full name. After change instead of  constructing complete module name from name and variant in dfu.py you will have to deconstruct variant name in `configurator_cli.py`

**Review by pdunaj — APPROVED** [review]

File: applications/nrf_desktop/doc/dfu.rst [hunk 11]
```diff
@@ -85,6 +85,7 @@ fwinfo
 
    * Version and length of the image.
    * Partition ID of the image, used to specify the image placement on the flash.
+   * Bootloader variant
```
Comment (MarekPieta) [inline applications/nrf_desktop/doc/dfu.rst:88]: From device perspective bootloader variant is a separate option. It's not part of the fwinfo. I already mentioned this in my previous comment

Comment (zycz) [reply applications/nrf_desktop/doc/dfu.rst:88]: Yes it is I added it to it
When you run fwinfo it will show you also bootloader variant

Comment (zycz) [reply applications/nrf_desktop/doc/dfu.rst:88]: However it is not in class fwinfo

Comment (MarekPieta) [reply applications/nrf_desktop/doc/dfu.rst:88]: It's description from perspective of an application module. Discussed offline.

File: scripts/hid_configurator/modules/dfu.py [hunk 12]
```diff
@@ -391,7 +414,9 @@ def dfu_start(dev, img_length, img_csum, offset):
     # verified by dfu sync at data exchange.
     event_data = struct.pack('<III', img_length, img_csum, offset)
 
-    success = dev.config_set('dfu', 'start', event_data)
+    dfu_module_name = dev.get_complete_module_name('dfu')
+    if dfu_module_name:
+        success = dev.config_set(dfu_module_name , 'start', event_data)
 
     if success:
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:421]: Variables refernced before assignment. Also in other places.
Causes runtime exceptions.

File: scripts/hid_configurator/modules/dfu.py [hunk 13]
```diff
@@ -357,7 +379,12 @@ def fwinfo(dev):
 
 
 def fwreboot(dev):
-    success, fetched_data = dev.config_get('dfu', 'reboot')
+    dfu_module_name = dev.get_complete_module_name('dfu')
+    if dfu_module_name:
+        success, fetched_data = dev.config_get(dfu_module_name , 'reboot')
+    else:
+        print('Module DFU not found')
+        return None
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:387]: Shouldn't we return `False, False` as `success, rebooted` pair? To keep values returned by the function consistent.

File: scripts/hid_configurator/modules/dfu.py [hunk 14]
```diff
@@ -391,7 +423,12 @@ def dfu_start(dev, img_length, img_csum, offset):
     # verified by dfu sync at data exchange.
     event_data = struct.pack('<III', img_length, img_csum, offset)
 
-    success = dev.config_set('dfu', 'start', event_data)
+    dfu_module_name = dev.get_complete_module_name('dfu')
+    if dfu_module_name:
+        success = dev.config_set(dfu_module_name , 'start', event_data)
+    else:
+        print('Module DFU not found')
+        return None
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:431]: return `False` ? The function used to return boolean indicating if operation was successful.

File: scripts/hid_configurator/modules/dfu.py [hunk 15]
```diff
@@ -543,7 +580,12 @@ def dfu_checkpoint(img_csum, img_length, offset):
 
         # Send data to the device
         logging.debug('Send DFU request: offset {}, size {}'.format(offset, chunk_len))
-        success = dev.config_set('dfu', 'data', chunk_data)
+        dfu_module_name = dev.get_complete_module_name('dfu')
+        if dfu_module_name:
+            success = dev.config_set(dfu_module_name , 'data', chunk_data)
+        else:
+            print('Module DFU not found')
+            return None
```
Comment (MarekPieta) [inline scripts/hid_configurator/modules/dfu.py:588]: Function returns pair: `success, offset`

**Review by MarekPieta — APPROVED** [review]
A few last comments regarding values returned on error.
Please keep in mind to update release notes (and state that the new implementation is not backwards compatible to avoid confusion among users)

File: doc/nrf/releases/release-notes-changelog.rst [hunk 16]
```diff
@@ -69,6 +69,13 @@ nRF9160: Asset Tracker v2
 
 * |no_changes_yet_note|
 
+nRF Desktop
+-----------
+
+* Added:
+
+  * Added bootloader variant as a DFU module variant for config channel communication. The new implementation is not backwards compatible.
```
Comment (pdunaj) [inline doc/nrf/releases/release-notes-changelog.rst:77]: > The new implementation is not backwards compatible.

Why? You don't need to check the variant. Everything else will be working as it was and old script should work without any issues.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 17]
```diff
@@ -123,6 +130,13 @@ Partition Manager
 
 * |no_changes_yet_note|
 
+HID Configurator
+----------------
+
+* Added:
+
+  * HID Configurator now recognizes bootloader variant as a DFU module variant for config channel communication. The new implementation is not backwards compatible.
```
Comment (pdunaj) [inline doc/nrf/releases/release-notes-changelog.rst:145]: > The new implementation is not backwards compatible.

Let's be specific and tell that since new version of script checks for variant it expects firmware to support this request.

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 18]
```diff
@@ -130,6 +137,13 @@ Partition Manager
...
```
Comment (greg-fer) [inline doc/nrf/releases/release-notes-changelog.rst:145]:
```suggestion
  * HID Configurator now recognizes the bootloader variant as a DFU module variant for the configuration channel communication.
  The new implementation is not backward compatible.
```

**Review by greg-fer — CHANGES_REQUESTED** [review]

**Review by greg-fer — APPROVED** [review]

**Review by pdunaj — APPROVED** [review]
